### ISSUE TITLE
Applicability of fast-cma-es algorithm MO-DE to multidimensional (>1000) problems

### ISSUE THREAD

**Issue by artiko88** [dietmarwo/fast-cma-es#27]
Hi and thanks for such a great library!

I am trying to train a neural network for a specific reinforcement learning environment using direct weights perturbation instead of graditn-based algorithms like PPO or SAC. Currenly, i've been using estool which is equipped with CMA-ES, OpenAI-ES and NES. CMA-ES shows some promise but it is hard for me to express fitness with a single value so i was looking for a multiobjective algorithms that can support constraints and train several objectives at the same time. Objective function is extremely expensive to evaluate (abnut 2 minutes per trial), but i am equipped with a multicore machine, so each population is around 96 individuals. 

I was thinking about using MODE algo from your library for this task. Do you think it is going to be appicable in terms of being able to train a plausible policy within sensible time? There are about 1000-3000 dimensions to optimize.

**Comment by dietmarwo** [comment]
Because evaluations are expensive, optimization strategies must be highly sample efficient. 
The goal is to make the best possible decision with as few evaluations as possible.

My favourite optimizer for this kind of tasks is https://github.com/optuna/optuna . 
It can do parallelization (as MODE) but is specialized for sample efficiency. 

Do you know https://github.com/google/evojax ? It aims at performing optimization 
directly on GPUs and has a Python interface. Some of the evojax algorithms are taken from fcmaes. 
It is useful if the objective itself can be implemented with JAX. 

>hard for me to express fitness with a single value

If there is any way to do that - using some weighted sum approach - this will speed up convergence. 

MODE with nsga_update = True (the default) is another option, but with 2 minutes per trial convergence will take some time. 
Differential evolution (the variant implemented in fcmaes) is another alternative if you can privide a single objective. 
It should work better than CMA-ES for your application. 

Think hard if there is any way to speed up the objective evaluation, this would be very helpful.

**Comment by artiko88** [comment]
> Because evaluations are expensive, optimization strategies must be highly sample efficient. The goal is to make the best possible decision with as few evaluations as possible.
> 
> My favourite optimizer for this kind of tasks is https://github.com/optuna/optuna . It can do parallelization (as MODE) but is specialized for sample efficiency.

I actually use Optuna extensively for other tasks and i was thinking that it is made more for ML hyperparameters optimization mainly and using it for direct optimization of NN weights is kind of a hacky way. But it has TPESampler which is capable of multiobjective optimization and i use it a lot. Maybe i should try it.

> Do you know https://github.com/google/evojax ? It aims at performing optimization directly on GPUs and has a Python interface. Some of the evojax algorithms are taken from fcmaes. It is useful if the objective itself can be implemented with JAX.

Yes, i do know evojax. Unfortunately, my objective function cannot (as of time) be transfered to GPU. Should be calculated on a CPU.

> > hard for me to express fitness with a single value
> 
> If there is any way to do that - using some weighted sum approach - this will speed up convergence.
> 
> MODE with nsga_update = True (the default) is another option, but with 2 minutes per trial convergence will take some time.

Any reason to hope that after 1000 generations with popsize of 100 something worthy can conjure up with MODE? I know that PGPE is used for hugh dimensional problems, how about MODE, is there any number of dimensions that it strives with?

> Differential evolution (the variant implemented in fcmaes) is another alternative if you can privide a single objective. It should work better than CMA-ES for your application.

So DE as implemented in fcmaes is usually faster than CMA-ES?

Interesting. CMA-ES shows some promise when it is trying to converge, but when components of the summed objective start to conflict, it kinda stangates and converges on a suboptimal solution. So i was thinking that providing constraints can speed up converence due to the fact that constraints will guide the search away from unfeasible solutions. 

> Think hard if there is any way to speed up the objective evaluation, this would be very helpful.

**Comment by dietmarwo** [comment]
>is made more for ML hyperparameters optimization 

Sample efficiency is most important for hyperparameters, but it is generally applicable for expensive objectives. 

>is there any number of dimensions that it strives with?

There is no limit, but convergence is slower with high dimensions. You can try a lower population size, but this limits your ability to parallelize and increases the chance to get stuck at local minima. The NSGA-II population update method MODE  uses is popular for a reason - to works quite well in most cases. CMA-ES works very good for space flight dynamics but is not generally preferable for all applications. https://github.com/dietmarwo/fast-cma-es/blob/master/fcmaes/pgpecpp.py which I implemented for EvoJax also can parallelize fitness evaluation for the whole population and may work - using a weighted sum for the objective. 

https://github.com/dietmarwo/fast-cma-es/blob/master/tutorials/Surrogate.adoc#car-design and https://github.com/dietmarwo/fast-cma-es/blob/master/tutorials/Water.adoc show applications where I successfully could speed up the evaluation of the objective compared to the original implementation.  Even the best algorithms struggle if confronted with a high dimensional and expensive problem. I always try to speed up the objective first. 

Did you check methods which work with derivatives? Many high dimensional problems can be solved with these if the solution landscape is not rugged. 

Another idea: Quality-Diversity https://github.com/dietmarwo/fast-cma-es/blob/master/tutorials/MapElites.adoc applied
to check the whole parameter space. You then may reduce the parameter bounds to investigate "interesting" areas using another method. But this also needs a lot of evaluations.

**Comment by artiko88** [comment]
Thank you, those are very good suggestions!

To give you more content about the objective function. Actually, i am trying to train a policy for a trading strategy that trades in an HFT setting. The strategy trades two highly correlated crypto assets. So it has several objectives: generate proft, generate profit with least amount variance, avoid drawdowns, generate large nuber of trades, stay in delta neutral position etc. Unfortunately, such strategy cannot be simulated using vectorized operations on numpy arrays - i have to use a backtester with full reconstruction of historical data, including full book reconstruction, simulating trades, latencies, slippage etc. The backtester is written in Rust with a Python interface for it, so it can already be considered fast enough in its own league. To get a fitness value i ran 10-15 runs on some preset historical periods, derive an average (or a min) value and return it as a fitness score. Decisions of the strategy are derived from a neural network policy, which is a fully connected MLP with two hidden layers (32x32 or larger). Actually, i do not know what NN setup is optimal for my task - it may be that even a single layer is enough, thus reducing the number of dimensions, or maybe a larger network is necessary. This has to be tested empirically.

Environment is wrapped in an OpenAI GYM Env, so that each step is returns a state value and an NN provides an optimal set of actions (a vector of 4 cont. values). 
The reason i turned to grad-free method is due to the fact that it is hard to derive a reward for such a noisy environment - the environment is very stochastic, rewards are delayed and noisy. I tried population based PPO where several agents are trained simulatensously and constantly evolve hyperparameters of the algo and even shape the reward function by some final fitness function, but decided to try grad-free method, since final fitness function is more important. Hell, even single fitness function does not descrbe the performance in enough details, so i am looking for an MO algo that can deal with the task. The hypothesis is that multiobjective algo will speed up convergence by directly using constraints, so the speed up will be gained by mapping unfeasible areas of search.

**Comment by dietmarwo** [comment]
Thank for sharing these details. This is a quite an ambitous project. 

>Actually, i do not know what NN setup is optimal for my task - it may be that even a single layer is enough, thus reducing the number of dimensions, >or maybe a larger network is necessary. This has to be tested empirically.

This should have high prio since a simpler net topology (less nodes / layers) speeds up evaluation significantly.  This looks like a typical hyperparameter optuna task. 

>So it has several objectives: generate proft, generate profit with least amount variance, avoid drawdowns, generate large nuber of trades, stay in >delta neutral position etc.

If a single evaluation is 2 minutes I still would try single objective optimization first, may be several runs using different objective weights. For such a high dimension MO will require more evaluations. But don't use CMA-ES. 

If you use fcmaes, try PGPE and CRFMNES. BiteOpt would also be good, but it cannot parallelize the evaluation of a population.

By the way, how does this work in your scenario? Do you use a whole cluster of graphic cards? Or do you evaluate the neural net using the CPU?

Your decision variables are the weights of the neural net ?

CMA-ES and even DE don't work very well with NN-weights - at least this is my experience with EvoJax. And I am sceptical about Mode/NSGA-II. 

PGPE and CRFMNES can use an initial guess.  You could try to find such a guess by simplyfying your Rust backtest avoiding some details,  so that it becomes faster and run the optimization with this simplified backtest first. 

>The hypothesis is that multiobjective algo will speed up convergence by directly using constraints, so the speed up will be gained by mapping >unfeasible areas of search.

I am sceptical. A multi objective algo priorizes elimination of constraint violations.  You can get a similar effect with a single objective adding a penalty for constraint violations. If for instance f(x) < c for an argument vector x and a constant c you could add something like

fx = f(x)
penalty = 0 if fx < c else offset + factor*(fx - c)

If f(x) is cheap but objective evaluation is expensive you could evaluate the constraint first and
evaluate the objective only if the constraint value is below some threshold. The offset makes
sure that arguments fulfilling constraits always get a better evaluation, if the offset is 
chosen larger than the "worst case" objective value. 

MODE cannot do this "out of the box" since it has no clue which offset would work.

**Comment by artiko88** [comment]
>Thank for sharing these details. This is a quite an ambitous project.

Oh, not more ambitiuos than what you have done in this repo! :0

>This should have high prio since a simpler net topology (less nodes / layers) speeds up evaluation significantly. This looks like a typical hyperparameter optuna task.

>Your decision variables are the weights of the neural net ?

Yes, looks like a task for Optuna, but in order to compare different architectures, one has to try to train a few via whole process of optimization. I cant see how i can use Optuna to simply find out the proper architecture first and then tune the weights with an ES/GA/DE algo. But the simpler the network, the less dimensions the problem becomes, the faster it can be trained - this is true, i agree.

>If f(x) is cheap but objective evaluation is expensive you could evaluate the constraint first and
evaluate the objective only if the constraint value is below some threshold. The offset makes
sure that arguments fulfilling constraits always get a better evaluation, if the offset is
chosen larger than the "worst case" objective value.

Do not quite get how  i can derive constraint values without running whole objective function. Or did i misunderstand you?

So i am gonna try CRFMNES, will tryto include constraints as penalties. Will also try to somehow lighten the simulation so that objective can be derived faster. Will try MOBO out of interest.

**Comment by dietmarwo** [comment]
>Do not quite get how i can derive constraint values without running whole objective function. Or did i misunderstand you?

That is only sometimes possible, probably not in your case. 

> is that multiobjective algo will speed up convergence by directly using constraints, so the speed up will be gained by mapping unfeasible areas of search.

But in that case a MO-algorithm cannot do wonders - you can do  the same - priorisation of feasible solutions - using a penalty with a single objective algo. MO-algorithms fit well, if you really need the the whole pareto front, not just a good
combination of objectives. Quality-Diversity algos collect even more solutions. You get huge number of "interesting" solutions and can choose after the optimization is completed applying any criteria. But the price is: You need more evaluations. Here is a CUDA based QD implementation https://qdax.readthedocs.io/en/latest/ . 

>So i am gonna try CRFMNES, will tryto include constraints as penalties. Will also try to somehow lighten the simulation so that objective can be derived faster. Will try MO-DE out of interest.

Sounds good.
